calyx/src/passes/clk_insertion.rs

```diff
@@ -28,18 +28,20 @@ impl Visitor for ClkInsertion {
     ) -> VisResult {
         let builder = ir::Builder::from(comp, sigs, false);
 
-        for cell_ref in &builder.component.cells {
+        let mut continuous_assignments =
+            std::mem::take(&mut builder.component.continuous_assignments);
```

Comment:
What does this line do? If we need to use this pattern generally, I'd rather have it abstracted over in some method than have to explain `mem::take` to people.

Reply:
So the problem in this particular file was that we wanted to iterate over the cells and push changes into the continuous assignments. Originally this was fine because the iteration took an immutable reference to the cells field and then a mutable reference was taken to the continuous assignments field from the mutable reference to the component.

Since cells is no longer directly accessible this means that iterating over it now takes an immutable reference to the whole component through the `iter_cells` method, rather than just the cells field on the component. This unfortunately makes it impossible to mutably access continuous_assignments through the component since it already has an immutable access being used to iterate. It took me a while to wrap my head around why the original code worked and why the change of iterating through a method didn't.

Anyway, to solve this we just needed to separate the continuous assignments from the component for the duration of the iteration, so all this line is doing is taking the current continuous assignments out of the component (temporarily replacing them with an empty vec within the component) and putting them in a mutable local variable so we can iterate through the component method and still make changes. At the end we reassign the (changed) continuous assignments back in to the component and erase the temporary vector.

We could abstract over this, though it would just be a single line invoking `mem::take` so I'm not sure that's much of a win. Is `mem::take` considered confusing in general?

Reply:
This highlights one of the downsides of these kinds of abstractions in rust atm. When you access field names directly, Rust can figure out that you are only mutating part of a structure and thus you can still have access to the rest of it. However, when you hide things behind methods Rust can no longer figure this out because there is no way to express this info in the type of a function.

That being said, abstraction is still nice. One possible solution is to have a function with a type something like:
```
fn iter_cells_cont_assgn_mut(&mut self, f: fn(&Cell, &mut Vec<cont assgn>) -> ())
```
so that internal to the function, Rust can work with the disjoint structure information. This has the downside of not letting you use a for loop, but I think that's ok.

Reply:
Personally I find that less intuitive than doing what I did, but I also wrote what I did so that's not saying much. Passing in a function type definitely makes sense though and would likely simplify things somewhat, so maybe that's the way to go.

Reply:
One other option would be to wrap the map construct for the cells and groups with a custom struct which exposes the methods that we directed through the component itself and make those fields public again. Downside is that it again reveals more about the internal representation that is not necessarily in-scope for people using the component, though I don't know how much of an issue that is